packages/renderer-commonmark/src/index.ts

```diff
@@ -103,114 +164,191 @@ export default class CommonmarkRenderer extends Renderer {
+   *
+   * ```js
+   * function () {}
+   * ```
    */
-  *'link'(props: { href: string }): IterableIterator<string> {
-    let [before, text, after] = yield* split();
-    return `${before}[${text}](${props.href})${after}`;
+  *'code'(props: { style: CodeStyle, info?: string }, state: State): IterableIterator<string> {
+    state.push({ isPreformatted: true, htmlSafe: false });
+    let text = yield;
+    state.pop();
+    let code = text.join('');
+
+    if (props.style === 'fence') {
+      code = '\n' + code;
+      let info = props.info || '';
+      let newlines = '\n';
+      if (state.get('isList') && state.get('nextAnnotation')) {
+        newlines += '\n';
+      }
+
+      if (code.indexOf('```') !== -1) {
+        return `~~~${info}${code}~~~${newlines}`;
+      } else {
+        return `\`\`\`${info}${code}\`\`\`${newlines}`;
+      }
+    } else if (props.style === 'block') {
+      return code.split('\n').map(line => `    ${line}`).join('\n') + '\n';
+    } else {
+      // Inline code must be at least 1 letter
+      if (code.length === 0) {
+        return '` `';
+      } else {
+        let backticks = '`'.repeat(getNumberOfRequiredBackticks(code));
+        return `${backticks}${code}${backticks}`;
+      }
+    }
+  }
+
+  *'html'(props: { type: string }, state: State): IterableIterator<string> {
+    state.push({ isPreformatted: true, htmlSafe: true });
+    let text = yield;
+    state.pop();
+    let html = text.join('');
+    if (props.type === 'block') {
+      return html + '\n';
+    }
+    return html;
   }
 
   /**
    * A list item is part of an ordered list or an unordered list.
    */
   *'list-item'(_, state: State): IterableIterator<string> {
-    let indent: string = '   '.repeat(state.get('indent'));
-    let rawItem: string[] = yield;
     let index: number = state.get('index');
-    let item: string = rawItem.join('').split('\n').map(line => indent + line).join('\n').trim();
-
-    if (state.get('type') === 'ordered-list') {
-      item = `${indent}${index}. ${item}`;
+    let delimiter = state.get('delimiter');
+    let marker: string = delimiter;
+    if (state.get('type') === 'numbered') {
+      marker = `${index}${delimiter}`;
       state.set('index', index + 1);
-    } else if (state.get('type') === 'unordered-list') {
-      item = `${indent}- ${item}`;
     }
+    let indent = ' '.repeat(marker.length + 1);
+    let text: string[] = yield;
+    let item: string = text.join('');
+    let firstCharacter = 0;
+    while (item[firstCharacter] === ' ') firstCharacter++;
+    let lines = item.split('\n');
+    lines.push(lines.pop().replace(/( )+$/, ''));
+    lines.unshift(lines.shift().replace(/^( )+/, ''));
+    let [first, ...rest] = lines;
 
-    return item;
+    item = ' '.repeat(firstCharacter) + first + '\n' + rest.map(line => indent + line).join('\n').replace(/( )+$/, '');
+    if (state.get('hasCodeBlockFollowing')) {
+      return ` ${marker}    ${item}`;
```

Comment:
Q: this implies a list item containing a code block? does this potentially change depending on inline/block/fence? (I guess i should have just read the tests first huh... trying to be thorough though)

Reply:
Ugh. This is what it is:

```
 -   hello
    I'm a code block _outside_ the list

Reply:
the extra newline in between doesn't terminate the list item / list?
man... yeah this one sucks! haha

Reply:
It doesn’t. 😭

Reply:
i just started playing around in the commonmark dingus and omg that's annoying. but i think maybe it's just that we should decide to always use fence blocked rather than "implicit" code blocks (via indentation). that would be a good output rule on our end to avoid ambiguity

Reply:
We will definitely do that.